TypeScript Conversion: media, units, react packages

swissspidy opened the pull request:
## Context

<!-- What do we want to achieve with this PR? Why did we write this code? -->

This is part of our ongoing effort to convert (parts of) our codebase to TypeScript. See #1715 for details.

## Summary

<!-- A brief description of what this PR does. -->

This marks the first step of migrating the `media` package to TypeScript, unblocking better typing usage in all consumer packages relying on it.

Due to my struggles with `Could not find a declaration file for module ...` errors, I also ended up converting the `units` and `react` packages (which are dependencies) to React.

Figured I'll keep everything in this one PR since it seemed so intertwined.

Success story:

* Important bug uncovered thanks to TypeScript, see https://github.com/GoogleForCreators/web-stories-wp/pull/12127#discussion_r949984838

## Relevant Technical Choices

<!-- Please describe your changes. -->

* Added custom typings for the `shallow-equal` module
* Added custom typings for globals like `WEB_STORIES_ENV`
* Remove custom `useReduction` fork in favor of the `use-reduction` package directly.
  Adapts Jest config to transpile `use-reduction` package because it's using ESM
* Removed `ResizeObserver` polyfill. Ran into https://github.com/que-etc/resize-observer-polyfill/issues/83 and figured we probably don't need this anymore anyway due to improved browser support.

## To-do

<!-- A list of things that need to be addressed in this PR or follow-up changes. -->

* It would be nice to eventually import a better `Element` type from a shared package (`elements` package?)
* Be able to move custom typings to the top-level `typings` folder
* Figure out better solution to address `Could not find a declaration file for module ...` errors so we can better convert packages one at a time.
* Figure out better typings for `UnitsProvider` and `useUnits` and `useContextSelector`
* Add some documentation for contributors

## User-facing changes

<!--
Please describe your changes.
Include before/after screenshots or a short video.
-->

N/A

## Testing Instructions

<!--
How can the changes in this PR be verified?
Please provide step-by-step instructions how to reproduce the issue, if applicable.
Write step-by-step test instructions aimed at non-tech-savvy users, even if the PR is not user-facing.
-->

<!-- ignore-task-list-start -->
- [x] This is a non-user-facing change and requires no QA
<!-- ignore-task-list-end -->

This PR can be tested by following these steps:

1.


## Reviews

### Does this PR have a security-related impact?

<!-- Examples: new APIs, changes to KSES, etc.  -->
No

### Does this PR change what data or activity we track or use?

<!-- Examples: changes to telemetry, new third-party APIs -->
No

### Does this PR have a legal-related impact?

<!-- Examples: new images with unknown sources, new production dependencies with incompatible licenses -->
No

## Checklist

<!-- Check these after PR creation -->

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

<!--
Please reference the issue(s) this PR addresses.
No URLs, just the issue numbers.
Use "Fixes #123" if it fixes an issue.

NOTE: One reference per line!

Example:

Fixes #123
Partially addresses #456
See #789
-->

Fixes #12087

swissspidy commented on packages/media/src/getCanvasBlob.ts:
```diff
@@ -14,11 +14,11 @@
     canvasEl.toBlob((blob) => {
-      resolve(blob, 'image/jpeg');
-    });
+      resolve(blob);
+    }, 'image/jpeg');
```
We've never spotted this bug before

swissspidy committed:
Convert media package to TypeScript

timarney commented on packages/media/src/types.ts:
```diff
@@ -200,6 +201,161 @@ export { ResourcePropTypes };
+  /** The optional author of the media object. */
+  author?: AttributionAuthor;
+  /** The optional url to register the media usage. */
+  registerUsageUrl: string;
```
Need to make this optional ya?

registerUsageUrl?: string;

swissspidy replied:
Yup indeed!

Now wondering whether `author` and `registerUsageUrl` should be split up into its own interface...

swissspidy committed:
Improve some types

swissspidy committed:
Try partially converting referenced packages

swissspidy commented on packages/media/src/getResourceSize.ts:
```diff
@@ -19,6 +19,14 @@
  */
 import { FULLBLEED_HEIGHT, PAGE_WIDTH } from '@googleforcreators/units';
```
So this `Could not find a declaration file for module '@googleforcreators/units'.` error has been nagging me for a while now and can't figure out why it's not working 🤔

miina replied:
Curious, what was the issue?

swissspidy replied:
That's the thing, I couldn't figure it out properly. Probably just some tsconfig misconfiguration somewhere, basically it did not like importing from this non-TS package. That's why I ended up just converting the `units` package to TypeScript 🙃 

For next time it would be great to resolve though. If you have any ideas, PLMK

swissspidy committed:
Make `registerUsageUrl` optional

miina commented on packages/units/src/dimensions.ts:
```diff
@@ -152,7 +184,11 @@ export function getBox(
+  element: Element,
+  pageWidth: number,
+  pageHeight: number
+) {
```
**💬 suggestion:**

This should probably also return `ElementBox`?

swissspidy replied:
Yes, but the return type is inferred from the function body itself, so it's technically redundant.

miina commented on packages/units/src/dimensions.ts:
```diff
@@ -109,14 +113,42 @@ export function editorToDataX(x, pageWidth, withRounding = true) {
+  y: number,
+  pageHeight: number,
+  withRounding = true
+) {
```
Here, too, return type?

miina commented on packages/media/src/types.ts:
```diff
@@ -200,6 +201,161 @@ export { ResourcePropTypes };
+  /** Whether the resource needs a CORS proxy. */
+  needsProxy: boolean;
+  /** Resource creation date. */
+  creationDate: string;
```
<!-- ideas that popped up from reviewing -->
**💡 thought:**
Should we also use `readonly` properties? E.g. `creationDate` sounds like one that shouldn't be rewritten.

(Edit: I'm seeing now elsewhere occasional usage of `readonly` as well)

swissspidy replied:
Probably yes, but that's easier to decide when going through the media-related code in the editor to see which ones can actually be readonly.

miina commented on packages/media/src/types.ts:
```diff
@@ -200,6 +201,161 @@ export { ResourcePropTypes };
+  /** Whether the resource has already been optimized. */
+  isOptimized: boolean;
+  /** Whether the resource is muted. */
+  isMuted: boolean;
```
<!-- potential concern, ask for clarification or investigation -->
**❓question:**
Wouldn't `isMuted` and other video-only props also be optional? 🤔

miina commented on packages/media/src/hasVideoGotAudio.ts:
```diff
@@ -14,15 +14,28 @@
+function hasVideoGotAudio(video: HTMLVideoElement): boolean {
   return (
-    video.mozHasAudio ||
+    (video.mozHasAudio as boolean) ||
```
<!-- potential concern, ask for clarification or investigation -->
**❓question:**
Does this need `as boolean` even though this prop's type is configured as `boolean` above? I thought it would be required only if TypeScript wouldn't know what the type is. Thoughts?

swissspidy replied:
Indeed

miina commented on packages/media/src/getTypeFromMime.ts:
```diff
@@ -14,14 +14,19 @@
-  return mimeType?.split('/')[0] || null;
-};
+function getTypeFromMime(mimeType: string): ResourceType {
+  return mimeType?.split('/')[0] as ResourceType;
```
**💬 suggestion:**

Maybe remove the `?` as well if the string is required?

miina commented on packages/media/src/getCanvasBlob.ts:
```diff
@@ -14,11 +14,11 @@
-      resolve(blob, 'image/jpeg');
-    });
+      resolve(blob);
+    }, 'image/jpeg');
```
<!-- highlight something positive -->
**👏 praise:**

Nice catch!

swissspidy replied:
Looks like this actually fixes #12132 too!

miina reviewed (commented):
Looking great!

swissspidy committed:
Convert rest of React package

Use use-reduction directly
Remove resize-observer-polyfill due to typing conflict, plus it should not be needed anymore anyway

swissspidy committed:
Add types for units package

swissspidy committed:
Update some more types

swissspidy committed:
Fix re-export

swissspidy committed:
Ignore dist-types folders for linting

swissspidy committed:
Merge branch 'main' into fix/12087-typescript-media

swissspidy committed:
Fix typing (hopefully)

swissspidy committed:
Jest: Transform use-reduction package

swissspidy committed:
Merge branch 'main' into fix/12087-typescript-media

swissspidy committed:
Merge branch 'main' into fix/12087-typescript-media

miina commented on packages/units/src/getBoundRect.ts:
```diff
@@ -130,7 +143,13 @@ export function getCorners(angle, x, y, width, height) {
   };
 }
 
-function getCorner(pivotX, pivotY, cornerX, cornerY, angle) {
```
<!-- propose improvements -->
**💬 suggestion:**
Looks like the utils in this file don't have return types, maybe add? Or is it purposefully so?

swissspidy replied:
They're implicit.

Right now we do not enforce explicit return types, though [there's a lint rule for that](https://typescript-eslint.io/rules/explicit-function-return-type/).

miina commented on packages/react/src/useContextSelector.ts:
```diff
@@ -53,8 +61,7 @@ function useContextSelector(context, selector, equalityFn = shallowEqual) {
   };
 
   // Update the selector fn to memoize the selected value by [equalityFn].
-  const patchedSelector = equalityFn ? equalityFnCallback : selector;
```
**💡 thought:**

Hm, if it would be called with `equalityFn = null`, would it fail now?

swissspidy replied:
Theoretically yes, though you'd probably get a type error. Plus I couldn't find any usage like that in our codebase.

miina commented on packages/media/src/types.ts:
```diff
@@ -200,6 +201,166 @@ export { ResourcePropTypes };
+  /** The resource's poster ID. */
+  posterId?: string;
+  /** Length in seconds. */
+  length: number;
```
<!-- trivial preference-based requests -->
**🥜 nitpick:**

I suppose `length` and `lengthFormatted` should also be optional since images don't have this?

miina reviewed (approved):
Looks good assuming the failing Karma tests are just flaky (re-running). Lots of files to go through so would be great if someone else would review as well.

swissspidy committed:
Make length optional

swissspidy committed:
Merge branch 'main' into fix/12087-typescript-media

swissspidy committed:
Merge branch 'main' into fix/12087-typescript-media

# Conflicts:
#	packages/media/src/getCanvasBlob.ts
